File: Concurrency/c++/the-dining-philosophers.cpp
#include <pthread.h> 
#include <semaphore.h> 
class DiningPhilosophers {
private:
    vector<sem_t> fork;
    sem_t mutex;
public:
    DiningPhilosophers() {
        fork.resize(5);
        sem_init(&mutex, 0, 1); 
        for(int i=0;i<5;i++)
            sem_init(&fork[i], 0, 1); 
    }
    
    ~DiningPhilosophers() {
        sem_destroy(&mutex); 
        for(int i=0;i<5;i++)
            sem_destroy(&fork[i]); 
    }

    void wantsToEat(int philosopher,
                    function<void()> pickLeftFork,
                    function<void()> pickRightFork,
                    function<void()> eat,
                    function<void()> putLeftFork,
                    function<void()> putRightFork) {
		sem_wait(&mutex); 
        
        sem_wait(&fork[philosopher]); 
        sem_wait(&fork[(philosopher+1)%5]); 
        
        sem_post(&mutex); 
        
        pickLeftFork();
        pickRightFork();
        eat();
        putLeftFork();
        putRightFork();
        
        sem_post(&fork[philosopher]); 
        sem_post(&fork[(philosopher+1)%5]); 
    }
};

File: Concurrency/Java/the-dining-philosophers.java
class DiningPhilosophers {
    private Semaphore[] fork;
    private Semaphore mutex;
    public DiningPhilosophers() {
        this.fork = new Semaphore[5];
        this.mutex = new Semaphore(1, true);
        for(int i=0;i<5;i++)
            this.fork[i] = new Semaphore(1, true);
    }

    // call the run() method of any runnable to execute its code
    public void wantsToEat(int philosopher,
                           Runnable pickLeftFork,
                           Runnable pickRightFork,
                           Runnable eat,
                           Runnable putLeftFork,
                           Runnable putRightFork) throws InterruptedException {
        
        mutex.acquire();
        
        fork[philosopher].acquire();
        fork[(philosopher+1)%5].acquire();
        
        mutex.release();
        
        pickLeftFork.run();
        pickRightFork.run();
        eat.run();
        putLeftFork.run();
        putRightFork.run();
        
        fork[philosopher].release();
        fork[(philosopher+1)%5].release();
    }
}
